src/odemis/gui/cont/temperature.py

```diff
@@ -0,0 +1,124 @@
+
+    def __init__(self, main_frame, thermostat, max_duration=5):
+        '''
+        main_frame: (wx) main xrc frame of the GUI
+        thermostat: (HwComponent) a component with the VA "temperature"
+        max_duration: (float) in seconds, the maximum duration
+        '''
+
+        if not hasattr(thermostat, "temperature"):
+            raise ValueError("Thermostat is missing the temperature VA")
+        self.thermostat = thermostat
+
+        self.temperature_label = main_frame.temperature_display
+        self._fg_color = self.temperature_label.GetForegroundColour()
+        self.temperature_label.Show()
+
+        self._time_last_warning = 0  # time of the last warning
+        self._time_last_speed_warning = 0
+
+        # temperature history is a deque of tuples (timestamp, temperature)
+        self.temperature_history = collections.deque([(time.time(), self.thermostat.temperature.value)])
+        self.max_duration = max_duration
+        self.thermostat.temperature.subscribe(self._updateTemperature)
+
+    def _updateTemperature(self, temperature):
+        """
+        Subscriber to the temperature VA
+        """
+        # add the new temperature to the history with a timestamp
+        self.temperature_history.append((time.time(), temperature))
+        wx.CallAfter(self.temperature_label.SetLabel, u"{temp:.2f}°C".format(temp=temperature))
+        
+        # clear out values of the history that are too old i.e. older than the max_duration
+        while len(self.temperature_history) > 0:
+            t, _ = self.temperature_history[0]
+            if t < time.time() - self.max_duration:
+                self.temperature_history.popleft()
+            else:
+                break
+
+        # Get boundaries
+        safe_range = self.thermostat.getMetadata().get(md.MD_FAV_POS_RANGE, None)
+        safe_speed = self.thermostat.getMetadata().get(md.MD_SAFE_SPEED_RANGE, None)
+
+        # Calculate a linear regression of the temperature history to determine the speed
+        l = list(zip(*self.temperature_history))  # convert [(time, temp), ...] to [[time, ...], [temp...]]
+        try:
+            speed, _ = numpy.polyfit(l[0], l[1], 1)
+        except:
+            logging.exception("Could not compute the rate of change of temperature.")
+            return
+
+        # logging.debug(u"Temperature: %f °C, speed is %f °C / s", temperature, speed)
+
+        # determine setting the warning
+        if safe_range is not None and not safe_range[0] <= temperature <= safe_range[1] :
+            # temperature is out of range
+            # change colour to red
+            wx.CallAfter(self.temperature_label.SetForegroundColour, gui.FG_COLOUR_ERROR)
+            text = u"Temperature {temp:.2f}°C is outside of the safe operating range!".format(temp=temperature)
+            logging.warning(text)
+            # display messagebox warning, but only once per minute
+            if time.time() - self._time_last_warning > 60:
```

Comment:
Please put the 60 into a constant named something like `MIN_PERIOD_WARNING`, at the top of the file.